Mathlib/GroupTheory/FreeGroup/ReducedWords.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2025 Bernhard Reinke. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amir Livne Bar-on, Bernhard Reinke
+-/
+import Mathlib.Data.List.Chain
+import Mathlib.GroupTheory.FreeGroup.Reduce
+
+/-!
+This file defines some extra lemmas for free groups, in particular about (cyclically) reduced words.
+
+## Main declarations
+
+* `FreeGroup.Red.reduced`: the predicate for reduced words
+
+-/
+open List
+
+universe u
+
+variable {α : Type u}
+namespace FreeGroup
+
+variable {L L₁ L₂ : List (α × Bool)}
+
+namespace Red
+
+/-- Predicate asserting that the word `L` admits no reduction steps, i.e., no two neighboring
+elements of the word cancel. -/
+@[to_additive "Predicate asserting the word `L` admits no reduction steps, i.e., no two neighboring
+elements of the word cancel."]
+def reduced (L : List (α × Bool)) : Prop := List.Chain' (fun a b => ¬(a.1 = b.1 ∧ (!a.2) = b.2)) L
+
+@[to_additive (attr := simp)]
+theorem reduced_nil : reduced ([] : List (α × Bool)) := List.chain'_nil
+
+@[to_additive (attr := simp)]
+theorem reduced_singleton {a : (α × Bool)} : reduced [a] := List.chain'_singleton a
+
+@[to_additive (attr := simp)]
+theorem reduced_cons {a b: (α × Bool)} :
+    reduced (a :: b :: L) ↔ ¬(a.1 = b.1 ∧ (!a.2) = b.2) ∧ reduced (b :: L) :=
+  List.chain'_cons
+
+@[to_additive]
+theorem not_step_reduced : reduced L₁ → ¬ Step L₁ L₂
```

Comment:
The style of this declaration is not in the usual mathlib style. I would personally go for:
```
@[to_additive]
theorem not_step_reduced (h : reduced L₁) : ¬ Step L₁ L₂ := fun step => by
  induction step
  simp [reduced] at h
```